### ISSUE TITLE
Big performance drop when using ensemble model over separate calls

### ISSUE THREAD

**Issue by jcuquemelle** [triton-inference-server/server#7650]
**Description**
We have an ensemble of 2 models chained together (description of models below).

Calling only the "preprocessing" model yields a max throughput of 21500 QPS @ 6 Cpu cores usage
Calling only the "inference" model yields a max throughput of 44000 QPS @ 6 Cpu + 0.7 Gpu usage
Calling the "ensemble" model yields a max throughput of less than 8500 QPS @ 10 Cpu + 0.7 Gpu usage

Why is there such a performance drop when using the ensemble ? given the figures for individual models, I'd expect the output of the ensemble to be around the same as the individual model with the less throughput i.e. 21500 QPS  

The server-side metrics don't show any sign of blocking step in the inference pipeline although client-side the latency increases over time: 
```
Request Rate: 8500 inference requests per seconds                                                                                       
  Pass [1] throughput: 8500.57 infer/sec. p75 latency: 3735 usec                                                                        
  Pass [2] throughput: 8499.32 infer/sec. p75 latency: 6115 usec
  Pass [3] throughput: 8500.86 infer/sec. p75 latency: 1298 usec
  Pass [4] throughput: 8282.7 infer/sec. p75 latency: 74570 usec
  Pass [5] throughput: 8384 infer/sec. p75 latency: 133184 usec
  Pass [6] throughput: 8425.3 infer/sec. p75 latency: 165502 usec
  Pass [7] throughput: 8255.55 infer/sec. p75 latency: 191599 usec
  Client: 
    Request count: 92706
    Throughput: 8356.54 infer/sec
    Avg client overhead: 18.98%
    p50 latency: 149491 usec
    p75 latency: 178315 usec
    p90 latency: 190544 usec
    p95 latency: 213366 usec
    p99 latency: 263968 usec
    Avg gRPC time: 153756 usec (marshal 5 usec + response wait 153751 usec + unmarshal 0 usec)
  Server: 
    Inference count: 92705
    Execution count: 92705
    Successful request count: 92705
    Avg request latency: 924 usec (overhead 265 usec + queue 90 usec + compute 569 usec)

  Composing models: 
  inference, version: 1
      Inference count: 92705
      Execution count: 73789
      Successful request count: 92705
      Avg request latency: 562 usec (overhead 107 usec + queue 67 usec + compute input 81 usec + compute infer 278 usec + compute output
 27 usec)

  preprocessing, version: 1
      Inference count: 92704
      Execution count: 87071
      Successful request count: 92704
      Avg request latency: 237 usec (overhead 33 usec + queue 23 usec + compute input 96 usec + compute infer 77 usec + compute output 7
 usec)
```

**Triton Information**
Running in py3 offical container, version 24.06, on a Kube pod with 16 Cpus and 32GB RAM.

**To Reproduce**
Composing models configurations: 
[ensemble.txt](https://github.com/user-attachments/files/17114778/ensemble.txt)
[preprocessing.txt](https://github.com/user-attachments/files/17114804/preprocessing.txt)
[inference.txt](https://github.com/user-attachments/files/17114824/inference.txt)

preprocessing: Onnx model on CPU, dynamic batching
Inference: Onnx model on Gpu with TRT execution provider, dynamic batching

**Expected behavior**
The output of the ensemble is around the same as the individual model with the less throughput i.e. 21500 QPS

**Comment by tanmayv25** [comment]
The performance degradation might be attributed to interference between the two composing models. From the report, it indeed looks like the requests are not getting blocked.

Can you try running model_analyzer for the ensemble to get the most performant model configuration for your use case?
https://github.com/triton-inference-server/model_analyzer/blob/main/docs/ensemble_quick_start.md

**Comment by yinggeh** [comment]
> Why is there such a performance drop when using the ensemble ? given the figures for individual models, I'd expect the output of the ensemble to be around the same as the individual model with the less throughput i.e. 21500 QPS
Hi @jcuquemelle. Why do you think the output of the ensemble to be around 21500 QPS? The `inference` model depends on the output of `preprocessing` so they are executed sequentially. The latency of ensemble model is the sum of both models, which will result a QPS lower than 21500. cc @tanmayv25

**Comment by caitengwei** [comment]
For a fairer comparison, you may need to compare:
1. The throughput when calling ‘preprocessing’ and ‘inference’ directly with the same QPS simultaneously.
2. The throughput when using an ensemble to invoke both models.

This way, case #1 can take into account the contention between the two models on the CPU.

**Comment by jcuquemelle** [comment]
Indeed, that's a good way to remove any uncertainty on this, I expect that was also the idea of:
>I can't find it back, but IIRC I came across someone having witnessed the issue even with an ensemble of 1 model, i.e. calling a single model through an ensemble had a significant performance hit against calling the model directly.

For now we put ensembles aside as we need to move forward on production cases so we reverted to aggregate both models in a single graph, but it is something we could come back to eventually.

**Comment by yinggeh** [comment]
Hi @jcuquemelle. I ran benchmarks with your provided models and it matches your observation. However, if you compare throughput of ensemble_model with throughtputs of COMBINED composing models, the difference is not that huge.

To find the maximum throughput, you may need to tune `--max-threads` argument in perf_analyzer (4, 8, 16, 32, 64 etc.) At some point you will find increasing threads have no benefit but only higher latency.

```
Triton version: 24.06
Memory: 32G
CPU: Intel® Core™ i9-13900KF
GPU: NVIDIA RTX A6000

Max throughput (infer/sec)
# Individual
preprocessing: 39995
inference: 86946
ensemble_model: 15794

# Combined
preprocessing: 24997
inference: 26085
```
cc @tanmayv25 @caitengwei

**Comment by jcuquemelle** [comment]
@yinggeh Thanks for your experiments, and glad to see we have 2 PR in progress on that topic:

- https://github.com/triton-inference-server/server/pull/8061
- https://github.com/triton-inference-server/core/pull/429

From what I see on your benchmark, I expect the 2 models are running on CPU right ? which would explain why the throughput evens out when they are used in a combined way, whereas individually the inference is twice as fast as the preproc. In our setup, the preproc is done on Cpu and the inference on Gpu, I don't expect as much interaction.
Even though, in this case the drop between models run independently and in ensemble (25k -> 15k, i.e 40%) is still quite significant. 

I'll resume with experimentation when the above-mentioned reviews are merged for sure :)

**Comment by yinggeh** [comment]
@jcuquemelle Not exactly. The benchmarks are from the same models provided in the internal ticket, which preprocessing uses CPU and inference uses GPU.

I have a few questions which are important.
1. What exactly is the expectation here?
Based on my study of this ticket, the throughput of `ensemble_model`can never reach the same throughput of `preprocessing` because of the extra ensemble scheduler overhead, end-to-end latency and how perf_analyzer works. Without increasing gRPC infer thread count, I manage to increase `ensemble_model` throughput on my machine from 16k QPS to 28k QPS, but still far from `preprocessing` model 40k QPS. With increasing gRPC infer thread count, `ensemble_model` throughput 51k QPS is close to `preprocessing` 57k QPS but I have some concern over non-ensemble models performance.
Is there a specific bottomline of `ensemble_model` throughput required in your production?
2. As far as I know, the models provided are different from the ones you use, and the bottlenecks in `ensemble_model` can be different. Can you provide me some `perf_analyzer` data on the original models (preprocessing, etc.)?
